atlasdb-commons/src/main/java/com/palantir/util/SafeShutdownRunner.java

```diff
@@ -0,0 +1,67 @@
+/*
+ * (c) Copyright 2020 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.util;
+
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+import com.palantir.common.concurrent.NamedThreadFactory;
+import com.palantir.common.concurrent.PTExecutors;
+import com.palantir.logsafe.exceptions.SafeRuntimeException;
+
+public class SafeShutdownRunner implements AutoCloseable {
+    private final ExecutorService executor = PTExecutors
+            .newFixedThreadPool(5, new NamedThreadFactory("safe-shutdown-runner", true));
```

Comment:
I guess this is fine. A service with tons of TransactionManagers could end up having a bunch of stuck threads though, especially if things are re-opened and closed (though if things hang uninterruptibly I guess there isn't much we can do).

Note: This does not actually create the threads until needed so most of the time you only create one thread even though we're making a fixed thread pool of 5

Reply:
I would almost say we want cached here? we want to close whatever we can.

Reply:
And we should make internal ski product notify us when it gets an exception from here so we can investigate immediately.

Reply:
Ah, you're right that if we run, e.g.

```
<thing that takes 30 seconds>
<fast task>
<thing that hangs>
```

then this keeps around 2 threads for no reason. Cached is probably fine in that the number of threads is at worst bounded by number of callbacks. If we wanted to have a hard limit we could also do a ThreadPoolExecutor with core 0 and max 5 - I think either is fine

Reply:
Agreed, cached is probably fine here

Reply:
Also, it's bounded by number of bad tasks, which I suspect is 1. Either way is fine